### PR TITLE
New feature for muting email and/or in-app notifications

### LINKED ISSUE #7492: Filter Mail notifications
Do you have any plans to support filtering the mail type notifications? We would like to filter which type of mail notifications should be send to the users, i.e only related to password recovery, email dataset contact and at most publish new dataset action. Our documentalists generates hundreds or even thousands of emails per day, which is annoying (Dataverse generates notification for every single operation to all users in Dataverse group).

### LINKED ISSUE #8487: API & Notifications
Hi
is there any way to retrieve notification details through API ?
I managed to get a notifications list (see below) but no info about datetime / dataset involved, etc.

Thanks

`{'status': 'OK', 'data': {'notifications': [{'id': 1604, 'type': 'PUBLISHEDDS'}, {'id': 1589, 'type': 'PUBLISHEDDS'}, {'id': 1579, 'type': 'PUBLISHEDDS'}, {'id': 1568, 'type': 'PUBLISHEDDS'}, {'id': 1559, 'type': 'PUBLISHEDDS'}, {'id': 1550, 'type': 'PUBLISHEDDS'}, {'id': 1542, 'type': 'CREATEDS'}, {'id': 1535, 'type': 'PUBLISHEDDS'}, {'id': 1521, 'type': 'ASSIGNROLE'}, {'id': 1463, 'type': 'CREATEACC'}]}`

### PR DESCRIPTION
**Pull request by ErykKul**
**What this PR does / why we need it**:
This new feature can be turned off by setting notification.showMuteOptions to "false" in Bundle.properties. By default it is enabled now. This feature allows turning off notifications (email or in-app) for user specified types. The types of notifications that can be turned off by users for themselves can be limited by using empty notification.typeDescription property, e.g., notification.typeDescription.CONFIRMEMAIL is turned off by default (also in Bundle.properties).

**Which issue(s) this PR closes**:

- Closes #7492
- Closes #8487

**Special notes for your reviewer**:
I have used flags system (bitwise comparisons) in order to avoid a join on db. Retrieving user query is called frequently, so this is done in order to keep performance acceptable.

**Suggestions on how to test this**:
Remember to set the Bundle properties. Login as a user and go to notifications, you should be able to mute certain notifications only for yourself.

**Does this PR introduce a user interface change? If mockups are available, please link/include them here**:
The notifications tab has a new feature. This could be turned off in order to keep the user interface unchanged.

**Is there a release notes update needed for this change?**:
Yes.

image

Alternative UI:

image

### CODE REVIEW

**Comment by DieuwertjeBloemen** [comment]
Hi all,

Eryk and I have been looking at the option to customize your notifications and emails as this is something we would want in our (KU Leuven RDR) installation and it’s something that has been mentioned before by others in the community.

Above, in Eryk’s screenshot, you can get a first idea of how notifications can be turned on or off by users in our current testing set-up.
We’ve been discussing it a bit more, and we wanted input from others on whether they should be able to truly customize it notification per notification or whether installations should provide users with some basic settings, or maybe even another set-up would be better. Here are some suggestions that we came up with for now: 

**Scenario 1:** Leave it all up to the users. They are able to mute each notification and email separately.

**Scenario 2:** Only allow users to choose between ‘all notifications’ and ‘essential notifications’ (e.g. access request notifications and API token expiration notifications), with the admin then being able to decide what ‘essential notifications’ are for the dataverse via the admin dashboard or via API.

**Scenario 3:** The feature is only manageable from the admin level and not up to the users individually. Users don’t have any choice and the admin decides for the entire dataverse what the notifications are that a user receives for datasets in that dataverse.

Please provide us with some input and maybe further suggestions as we continue to work on this feature to make it a bit more user friendly.

Dieuwertje

**Comment by ErykKul** [comment]
This code change is between scenario 1 and 2: the admin can choose which notifications are essential and are not selectable by users, and the user can mute the remaining (not essential) notifications all at once, or one by one by picking them from the list.

**Comment by ErykKul** [comment]
I have added the support for the third scenario: you can set `notification.alwaysMuted` in bundle.properties to permanently mute certain notification types, where users cannot unmute them, e.g. `notification.alwaysMuted=ASSIGNROLE, REVOKEROLE`

File: src/main/java/edu/harvard/iq/dataverse/UserNotification.java
Comment (poikilotherm) [inline src/main/java/edu/harvard/iq/dataverse/UserNotification.java:50]: While I generally like the idea of using a custom made serialization of using a `long` for sth similar to a bitfield, this is more C than Java.

How about using `Set<Type>` directly? (Exposing an `EnumSet`) You could use `@PrePersist` and `@PostLoad` if you still want to serialize into a single database field, but make it encapsuled and transparent to users of the model class.

File: src/main/java/edu/harvard/iq/dataverse/authorization/providers/builtin/DataverseUserPage.java
Comment (poikilotherm) [inline src/main/java/edu/harvard/iq/dataverse/authorization/providers/builtin/DataverseUserPage.java:170]: Not sure we should create just another configuration mechanism...

Please have a look at the `SettingsWrapper` already present in this class, it wraps the `SettingsServiceBean` for JSF pages.

**Review by poikilotherm — CHANGES_REQUESTED** [review]
Thank you @ErykKul and @DieuwertjeBloemen for your work on this.

To make it more likely this code will be accepted upstream, you should try to discuss your configuration approach with the other devs.

File: src/main/java/edu/harvard/iq/dataverse/userdata/UserUtil.java
Comment (ErykKul) [reply src/main/java/edu/harvard/iq/dataverse/userdata/UserUtil.java:8]: Thank you for spotting this one out. I have just removed it. I am used to go programming and the strings package there and it got mixed up.

File: src/main/java/edu/harvard/iq/dataverse/UserNotification.java
Comment (ErykKul) [reply src/main/java/edu/harvard/iq/dataverse/UserNotification.java:50]: I have implemented your suggestion. I am not entirely sure if I did it exactly as you suggested, I am little bit concerned with the setters in the `AuthenticatedUser`. I still have a setter for the long value since it is used by a query. Since I needed to keep both types of setters (for a `Set` and a `long`), I need to keep the values in sync. Would you mind to look at this again? Thank you.

Comment (poikilotherm) [reply src/main/java/edu/harvard/iq/dataverse/UserNotification.java:50]: I would, but could you please push your commits? Thx :smile:

Comment (ErykKul) [reply src/main/java/edu/harvard/iq/dataverse/UserNotification.java:56]: Sorry :blush:

Comment (ErykKul) [reply src/main/java/edu/harvard/iq/dataverse/UserNotification.java]: I had some problems with git, lost my changes.... Finally it is working again. The configuration is much better now with setting wrapper.

**Review by poikilotherm — CHANGES_REQUESTED** [review]
Hey @ErykKul I left a few comments... :wink: 

What looks like isn't covered currently: what about concurring admin settings? What happens if a notification is in both never and always muted? I think it would be fine to define a "wining" type, but that should be clearly formulated/documented in code and IMHO also be logged, as this might cause confusion.

File: src/main/java/edu/harvard/iq/dataverse/settings/SettingsServiceBean.java
Comment (ErykKul) [reply src/main/java/edu/harvard/iq/dataverse/settings/SettingsServiceBean.java:514]: I have corrected the example. When processing the list I trim the entries from the whitespaces, so the previous example would have worked. I have added extra documentation explaining this.

File: src/main/java/edu/harvard/iq/dataverse/UserNotificationServiceBean.java
Comment (ErykKul) [reply src/main/java/edu/harvard/iq/dataverse/UserNotificationServiceBean.java:41]: OK, refactored.

File: src/main/java/propertyFiles/Bundle.properties
Comment (ErykKul) [reply src/main/java/propertyFiles/Bundle.properties:268]: No. I have just changed it.

File: src/main/java/edu/harvard/iq/dataverse/authorization/users/AuthenticatedUser.java
Comment (ErykKul) [reply src/main/java/edu/harvard/iq/dataverse/authorization/users/AuthenticatedUser.java:538]: Good suggestion, thank you.

File: src/main/java/edu/harvard/iq/dataverse/userdata/UserUtil.java
Comment (ErykKul) [reply src/main/java/edu/harvard/iq/dataverse/userdata/UserUtil.java:85]: For now I just used the new static conversion to String as in the new JPA implementation, where the muted emails and notifications columns are strings now. This more consistent with the remaining code IMO.

Comment (ErykKul) [reply src/main/java/edu/harvard/iq/dataverse/userdata/UserUtil.java:85]: I have used the JsonArrayBuilder, it is cleaner that way.

File: src/main/java/edu/harvard/iq/dataverse/authorization/users/AuthenticatedUser.java
Comment (ErykKul) [reply src/main/java/edu/harvard/iq/dataverse/authorization/users/AuthenticatedUser.java:130]: I have used String here and VARCHAR(1000) as db type.

File: src/main/java/edu/harvard/iq/dataverse/authorization/users/AuthenticatedUser.java
Comment (ErykKul) [reply src/main/java/edu/harvard/iq/dataverse/authorization/users/AuthenticatedUser.java]: I have made extra print and parse methods (and tests) specifically for enums. I use then `entry.name()` en `Enum.valueOf(clazz, name)`:
```java
    public static <E extends Enum> JsonArrayBuilder enumsToJson(Collection<E> collection) {
        JsonArrayBuilder arr = Json.createArrayBuilder();
        for (E entry : collection) {
            arr.add(entry.name());
        }
        return arr;
    }


    public static <E extends Enum<E>> List<E> parseEnumsFromArray(JsonArray enumsArray, Class<E> clazz) throws JsonParseException {
        final List<E> enums = new LinkedList<>();

        for (String name : enumsArray.getValuesAs(JsonString::getString)) {
            enums.add(Enum.valueOf(clazz, name));
        }
        return enums;
    }
```

File: src/main/webapp/dataverseuser.xhtml
Comment (ErykKul) [reply src/main/webapp/dataverseuser.xhtml:387]: @DieuwertjeBloemen

**Comment by ErykKul** [comment]
@poikilotherm, I have implemented the requested changes. Thank you for your detailed review, I appreciate it since I can learn some things early on. As for the user interface change, I think it should be first discussed by people responsible for functionality, e.g., @DieuwertjeBloemen.

I have also documented that the always muting has priority over never muting, if both are set for a certain type, a warning is printed in the log.

**Comment by DieuwertjeBloemen** [comment]
Hi,

I'm currently just collecting as much input as possible for the user interface. So, thanks for the input and I'll note it down. Then, I'll collect all input and work out a couple options for Eryk to look at to see what's technically feasable as a first version of the feature.
So, keep the user interface input coming (it's also welcome in [the google community post)](https://groups.google.com/g/dataverse-community/c/OoYf8OlJDsU)

**Comment by poikilotherm** [comment]
> @poikilotherm, I have implemented the requested changes. Thank you for your detailed review, I appreciate it since I can learn some things early on. 

Glad I was helpful :smile: 

> As for the user interface change, I think it should be first discussed by people responsible for functionality, e.g., @DieuwertjeBloemen.

Please note that as this is UI related, @TaniaSchlatter and maybe others from IQSS UI team will need to give their approval. It's good that you already provided some screenshots to look at.

If you find my suggestion useful, maybe you can add some screenshots of that variant, too.

> I have also documented that the always muting has priority over never muting, if both are set for a certain type, a warning is printed in the log.

Great!

It would be very nice to have some docs included in the user guide about this (as part of the PR), but that could wait until you have feedback by IQSS...

**Comment by ErykKul** [comment]
I have tried the selectManyCheckbox and I like it better like this, so I have committed these changes. If we decide for something else, then I can change it again. I have added a screenshot in the PR request.

**Comment by pdurbin** [comment]
> Please note that as this is UI related, @TaniaSchlatter and maybe others from IQSS UI team will need to give their approval. It's good that you already provided some screenshots to look at.

Absolutely. The other day I added this pull request into our project board at https://github.com/orgs/IQSS/projects/2 and automation moved it into the "Review" column, as expected. However, when we discussed this pull request during standup we decided that due to the UI/UX impact we aren't ready to simply review at a code level and move it to QA. So we put the pull request in a column called "Needs Discussion". I'm not sure when the right people will look at it and give feedback and I definitely don't want to squelch the awesome work and discussion I see going into this. I just say all this to try to set expectations since this isn't a simple bug fix. It's a feature that needs some design and consideration. All that said, please feel free to continue hacking and discussing (including in the lively thread at https://groups.google.com/g/dataverse-community/c/OoYf8OlJDsU/m/kxhm7r2WAgAJ ). Thanks!

**Comment by DieuwertjeBloemen** [comment]
Thanks for all the input. Currently, Eryk has some time off during Easter break, so he won't be working on it for the next two weeks. I'll now just collect all the input and work out a couple options for Eryk to look at to see what's technically feasable and then contact @TaniaSchlatter to discuss the UI and what we can or can't do in that regard.

**Comment by pdurbin** [comment]
Just a quick update that we talked about this pull request at tech hours this afternoon. Not in any depth, mind you, just that we know we owe you some feedback.

One concern we touched on today has already been voiced in the thread on the list:

> For curated Dataverse installations/collections, I think we need to make sure that users cannot opt out of important curation-related messages, e.g., the notification about a dataset having been submitted for review or sent back to author.

Also, I noticed there are merge conflicts but no rush to fix them until we've gotten back to you. Thanks for your patience.

**Comment by DieuwertjeBloemen** [comment]
Hi all,

Eryk and I ran through all the feedback and have some comments and further questions:

1. An admin dashboard doesn’t seem like something we want to focus our attention to at this moment. So, as a first iteration of the feature we would suggest making the settings manageable via API. This will ensure that the basic functionality is available and that our set-up doesn’t become too complicated from the get-go and that we can focus on first making the notifications and emails configurable. Is this something you’re all okay with?
2. We thought about the option to make the settings manageable on a dataverse collection. However, this would make the UI and overall feature a lot more complicated. For example, what do you do with users of multiple collections in a dataverse installation? So, as a basic release, we don’t want to include this option as it’s also not something we would use ourselves and would require more resources.

So, what we currently have as a concept is the following:

_The admin can set a series of 'essential' notifications/mails that users can't individually turn off and set notifications as ‘always muted’ that users can’t turn on. You can then set the other notifications/mails as something the users themselves can configure in the notifications tab._
And decause the admin can decide the settings for a set of notifications/mails to not be turned on or off by the user, the curation-related messages can very easily be protected from being turned off by users.

Please let me know if you have any questions or comments. Once we have the okay to continue with this set-up, we'll contact Tania to further work out the UI.

Kind regards,

Dieuwertje

**Comment by pdurbin** [comment]
@DieuwertjeBloemen generally speaking, yes, we're happy to implement some functionality in the API first.

In standup this morning and in Slack last week we talked about how a demo might be nice. Is this something you'd be willing to do? If so, would you also be able to organize it in terms of sending out a poll for scheduling it? Please let us know what you think. Thanks!

**Comment by DieuwertjeBloemen** [comment]
@pdurbin, We would definitely be open to doing a demo, that would also be a great way to get some input on what we have right now. We've never done one before, so should I just send out a poll for scheduling to the community via GitHub and the Google Community or somewhere else? Some of our team also usually attend the second community call, so we could also do it during one of those, if that would be easier.

**Comment by scolapasta** [comment]
Hi @DieuwertjeBloemen @ErykKul I went ahead and movedthis community dev while you work on the changes that were discussed on the community call. Once it's ready for review again, please let us know and we can move to Next Sprint for consideration. (it should be able to be fast tracked at that point, I hope)

**Comment by ErykKul** [comment]
Hi @scolapasta 
I have just implemented the only remark we have received, so we are ready on our side. From now on only the never muted notifications are grayed out, the always muted notifications are simply never shown (the users do not need to see notification types that they cannot receive and cannot unmute):

image

I think you can move i to Next Sprint.

**Comment by pdurbin** [comment]
@ErykKul can you please also add docs and a release note? From looking at your past pull requests, I'm not sure we've asking this of you before.

At minimum, the database settings you added need to be documented. They go here (at the bottom is fine): https://guides.dataverse.org/en/5.10.1/installation/config.html#database-settings

You are also very welcome to write up something in the Admin Guide. I'd say this fits neatly enough on this page: https://guides.dataverse.org/en/5.10.1/admin/user-administration.html

Finally, please create a short file to explain what you think should go in the release notes. The process is described here: https://guides.dataverse.org/en/5.10.1/developers/making-releases.html#write-release-notes

Thanks! Also, great demo! For anyone reading this that missed it, you can find it here:

- demo: https://harvard.zoom.us/rec/share/sFlpmpsfBOFyGHUiWcrLpLwnjpAa_CaHkV6WU1mBB56cUDXJ8hWUF50qcNIIQ_yV.alzT560S9diRN1TT
- slides: https://kuleuven-my.sharepoint.com/:p:/g/personal/dieuwertje_bloemen_kuleuven_be/EY4OnkcGx39BhA6vLWpizmEBC71P4MfgKfg2hF63S_sD7A

**Comment by ErykKul** [comment]
@pdurbin I have added the documentation. Can you review it? I was not sure where this should go, I have left it at the bottom of config.rst:
`.. To edit, use dvBrandingCustBlocks.drawio with https://app.diagrams.net
.. |dvPageBlocks| image:: ./img/dvBrandingCustBlocks.png
   :class: img-responsive`

**Comment by TaniaSchlatter** [comment]
@scolapasta and I reviewed this with @sbarbosadataverse, and it has her support.

**Comment by pdurbin** [comment]
@ErykKul sure, please look at some suggested changes at https://github.com/ErykKul/dataverse/pull/1

I agree with you about the dvBrandingCustBlocks stuff and fixed it (moving it with the image itself).

**Comment by ErykKul** [comment]
@pdurbin, thank you for your review. @DieuwertjeBloemen will take a look at the user guide.

**Comment by ErykKul** [comment]
@pdurbin, we have updated the notifications' user guide. Does it look OK now?

**Comment by scolapasta** [comment]
@ErykKul @pdurbin Correct, the next sprint is currently planned to start this Wednesday and I'm definitely planning on moving this one over.

File: src/main/java/propertyFiles/Bundle.properties
Comment (pdurbin) [inline src/main/java/propertyFiles/Bundle.properties:240]:
```suggestion
notification.muteOptions=Notification settings
```

I'm not in love with "Notification and mailing settings". (I'll put screenshots below for context.) "Notification settings" is shorter and can encompass both email and in-app varieties.

Screen Shot 2022-05-16 at 4 33 20 PM


Screen Shot 2022-05-16 at 4 25 39 PM

File: doc/sphinx-guides/source/user/account.rst
Comment (pdurbin) [inline doc/sphinx-guides/source/user/account.rst:165]: It would be nice if users could view and modify their notification settings via API.

This could, of course, be a future pull request, but if there's interest in adding this now, here's where the notifications API has been stubbed out: https://github.com/IQSS/dataverse/blob/v5.10.1/src/main/java/edu/harvard/iq/dataverse/api/Notifications.java

This issue is related:

- #8487

Having this API would also help with automated testing! 😄

What do you think? Again, it's fine to say that now's not the time. 😄

File: src/test/java/edu/harvard/iq/dataverse/authorization/users/AuthenticatedUserTest.java
Comment (pdurbin) [inline src/test/java/edu/harvard/iq/dataverse/authorization/users/AuthenticatedUserTest.java:410]: I added this test (and the one below). Hope it helps! Please feel free to expand on them if you like.

**Review by pdurbin — CHANGES_REQUESTED** [review]
@ErykKul I finally ran the code and poked around a bit. Overall, this looks great!

I did notice a few things I think we should change. Please see my comments in the review itself.

File: src/main/java/edu/harvard/iq/dataverse/util/json/JsonParser.java
Comment (ErykKul) [reply src/main/java/edu/harvard/iq/dataverse/util/json/JsonParser.java:294]: Here I went with ``enumClass``, you need to pass the Class of the ``enum`` you are processing.

File: src/main/resources/db/migration/V5.10.1.1__7492-muting-notifications.sql
Comment (ErykKul) [reply src/main/resources/db/migration/V5.10.1.1__7492-muting-notifications.sql:5]: I have changed it to text. I have also updated the bean:
`    @Column(columnDefinition="TEXT", nullable=true)
    private String mutedEmails;

    @Column(columnDefinition="TEXT", nullable=true)
    private String mutedNotifications;`

I hope I am not forgetting to something and that it is OK now...

File: src/main/java/propertyFiles/Bundle.properties
Comment (ErykKul) [reply src/main/java/propertyFiles/Bundle.properties:247]: I have changed it to "Dataverse collection is created". I think the only reason to leave it unchanged at this moment is if this is confusing to the user. Furthermore, I prefer to change it now than risk forgetting it.

File: src/test/java/edu/harvard/iq/dataverse/authorization/users/AuthenticatedUserTest.java
Comment (ErykKul) [reply src/test/java/edu/harvard/iq/dataverse/authorization/users/AuthenticatedUserTest.java:410]: Thanks, it looks good.

**Review by pdurbin — APPROVED** [review]
I've run the code and played with it but haven't done extensive testing. I did try to improve docs and tests here and there. Also, please note than in addition to the original issue (#7492), thanks to @ErykKul this pull request now also addresses #8487.

Also, as discussed in Slack, we renamed some SQL scripts in ea342ab so developers will need to fix up their flyway history (or drop their database and get set up again): https://guides.dataverse.org/en/5.10.1/developers/sql-upgrade-scripts.html#renaming-sql-upgrade-scripts

**Review by pdurbin — APPROVED** [review]

**Comment by kcondon** [comment]
@ErykKul I'm having trouble deploying this pr for testing as it is colliding on the flyway script number: 5.10.1.0.1 | 8533-semantic-updates. Would you mind updating your script number to avoid this? Thanks

**Comment by pdurbin** [comment]
@ErykKul @kcondon I renamed the flyway scripts in ea342ab. I'll take a look.

**Comment by kcondon** [comment]
In the API Guide, some of the notifications versus email endpoints have the same title, which makes it somewhat confusing:
Unmute In-app Notification by User is changed to Unmute Email Notification by User when referring to the email endpoint but for the others, it seems both notifications and emails use Notification in the title: Get All Muted In-app Notifications by User, Mute In-app Notification by User, Unmute In-app Notification by User

**Comment by kcondon** [comment]
@ErykKul I've finished my testing, thanks for all the work. I did find one thing that could use a look: when putting or deleting an end user notification type, eg. CREATEDS of either notification or email, that is an invalid string, eg. xCREATEDS, the return is {} and 500 error and the server log shows an error: [2022-05-23T18:36:32.514+0000] [Payara 5.2021.5] [WARNING] [AS-EJB-00056] [javax.enterprise.ejb.container] [tid: _ThreadID=75 _ThreadName=http-thread-pool::http-listener-1(1)] [timeMillis: 1653330992514] [levelValue: 900] [[
 A system exception occurred during an invocation on EJB Notifications, method: public [javax.ws.rs](http://javax.ws.rs/).core.Response [edu.harvard.iq](http://edu.harvard.iq/).dataverse.api.Notifications.muteEmailsForUser(java.lang.String)]]
[2022-05-23T18:36:32.514+0000] [Payara 5.2021.5] [WARNING] [] [javax.enterprise.ejb.container] [tid: _ThreadID=75 _ThreadName=http-thread-pool::http-listener-1(1)] [timeMillis: 1653330992514] [levelValue: 900] [[
javax.ejb.EJBException: No enum constant [edu.harvard.iq](http://edu.harvard.iq/).dataverse.UserNotification.Type.xCREATEDS in the logs

**Comment by ErykKul** [comment]
@kcondon I have improved the error handling. Thank you for testing and feedback!

**Comment by kcondon** [comment]
@ErykKul Thanks, that looks good.

**Comment by kcondon** [comment]
@ErykKul I've discussed the session issue with the dev team and it was viewed as minor impact and likely not easy to fix so we'll merge it. Thanks!

**Pull request merged**
